File: civo-navigate-demos/extism/extism-vert-x/src/main/java/garden/bots/starter/MainVerticle.java
package garden.bots.starter;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Promise;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServer;
import io.vertx.ext.web.Route;
import io.vertx.ext.web.Router;

import org.extism.sdk.Context;
import org.extism.sdk.manifest.Manifest;
import org.extism.sdk.wasm.WasmSourceResolver;

import java.nio.file.Path;

public class MainVerticle extends AbstractVerticle {

  @Override
  public void start(Promise<Void> startPromise) throws Exception {

    var resolver = new WasmSourceResolver();
    var wasmFilePath = System.getenv("WASM_FILE");
    var manifest = new Manifest(resolver.resolve(Path.of(wasmFilePath)));

    HttpServer server = vertx.createHttpServer();

    Router router = Router.router(vertx);
    Route route = router.route().method(HttpMethod.POST);

    route.handler(ctx -> {

      // This handler will be called for any POST request
      ctx.request().body().andThen(asyncRes -> {
        var name = asyncRes.result().toString();

        try (var extismCtx = new Context(); var plugin = extismCtx.newPlugin(manifest, true))
        {
          var output = plugin.call(System.getenv("FUNCTION_NAME"), name);
          
          ctx.response()
          .putHeader("content-type", "application/json; charset=utf-8")
          .end(output);
  
        }
        catch (Exception e) {
          ctx.response()
          .putHeader("content-type", "application/json; charset=utf-8")
          .end(e.getMessage());
        } 

      });

    });

    var httpPort = System.getenv("HTTP_PORT");

    server.requestHandler(router).listen(Integer.parseInt(httpPort), http -> {
      if (http.succeeded()) {
        startPromise.complete();
        System.out.println("HTTP server started on port " + httpPort);
      } else {
        startPromise.fail(http.cause());
      }
    });
    

    // curl -X POST -d 'Bob Morane' http://localhost:8888

  }
}

File: samples/wasm-runner/src/main/kotlin/garden/bots/starter/MainVerticle.kt
package garden.bots.starter

import io.vertx.core.AbstractVerticle
import io.vertx.core.Promise

import org.extism.sdk.Context
import org.extism.sdk.manifest.Manifest
import org.extism.sdk.wasm.WasmSourceResolver
import java.nio.file.Path
import kotlin.system.exitProcess
import kotlin.io.println


class MainVerticle : AbstractVerticle() {

  override fun start(startPromise: Promise<Void>) {

    val initializePlugIn = { path: String  ->
      runCatching { 
        val resolver = WasmSourceResolver()
        val manifest = Manifest(resolver.resolve(Path.of(path)))
        val ctx = Context() 
        ctx.newPlugin(manifest, true)
      }
    }

    initializePlugIn("./hello-world/hello-world.wasm")
      .onFailure { exitProcess(0) }
      .onSuccess {
        vertx
        .createHttpServer()
        .requestHandler { req -> 
          runCatching { 
            val output = it.call("helloWorld", "Bob Morane")
            req.response()
            .setStatusCode(200)
            .putHeader("content-type", "text/plain; charset=utf-8")
            .end(output)

          }.recoverCatching { 
            req.response()
            .setStatusCode(500)
            .putHeader("content-type", "text/plain; charset=utf-8")
            .end("Ouch 😡")
          }
        }
        .listen(8888) { http ->
          if (http.succeeded()) {
            startPromise.complete()
            println("HTTP server started on port 8888")
          } else {
            startPromise.fail(http.cause());
          }
        }
      }
  }
}
